File: zig/src/data_structures/queue.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// FIFO queue implementation using a ring buffer.
/// Time: O(1) amortized for all operations
/// Space: O(n)
pub fn Queue(comptime T: type) type {
    return struct {
        const Self = @This();

        items: std.ArrayList(T),
        head: usize,

        pub fn init(allocator: Allocator) Self {
            return .{
                .items = std.ArrayList(T).init(allocator),
                .head = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            self.items.deinit();
        }

        pub fn enqueue(self: *Self, value: T) !void {
            try self.items.append(value);
        }

        pub fn dequeue(self: *Self) ?T {
            if (self.head >= self.items.items.len) return null;
            const value = self.items.items[self.head];
            self.head += 1;
            // Compact when head reaches half the capacity
            if (self.head > self.items.items.len / 2 and self.head > 16) {
                const remaining = self.items.items.len - self.head;
                std.mem.copyForwards(T, self.items.items[0..remaining], self.items.items[self.head..]);
                self.items.shrinkRetainingCapacity(remaining);
                self.head = 0;
            }
            return value;
        }

        pub fn peek(self: *const Self) ?T {
            if (self.head >= self.items.items.len) return null;
            return self.items.items[self.head];
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.head >= self.items.items.len;
        }

        pub fn size(self: *const Self) usize {
            return self.items.items.len - self.head;
        }

        pub fn clear(self: *Self) void {
            self.items.clearRetainingCapacity();
            self.head = 0;
        }
    };
}

test "Queue operations" {
    const allocator = std.testing.allocator;
    var queue = Queue(i32).init(allocator);
    defer queue.deinit();

    try queue.enqueue(1);
    try queue.enqueue(2);
    try queue.enqueue(3);

    try std.testing.expectEqual(@as(?i32, 1), queue.dequeue());
    try std.testing.expectEqual(@as(?i32, 2), queue.dequeue());
    try std.testing.expectEqual(@as(?i32, 3), queue.dequeue());
    try std.testing.expectEqual(@as(?i32, null), queue.dequeue());
}

File: rust/src/data_structures/queue.rs
use std::collections::VecDeque;

/// A generic queue implementation using VecDeque.
/// Time: O(1) for enqueue, dequeue, peek
/// Space: O(n)
#[derive(Debug, Clone)]
pub struct Queue<T> {
    data: VecDeque<T>,
}

impl<T> Queue<T> {
    pub fn new() -> Self {
        Queue {
            data: VecDeque::new(),
        }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Queue {
            data: VecDeque::with_capacity(capacity),
        }
    }

    pub fn enqueue(&mut self, value: T) {
        self.data.push_back(value);
    }

    pub fn dequeue(&mut self) -> Option<T> {
        self.data.pop_front()
    }

    pub fn peek(&self) -> Option<&T> {
        self.data.front()
    }

    pub fn peek_mut(&mut self) -> Option<&mut T> {
        self.data.front_mut()
    }

    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    pub fn len(&self) -> usize {
        self.data.len()
    }

    pub fn clear(&mut self) {
        self.data.clear();
    }
}

impl<T> Default for Queue<T> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_queue() {
        let queue: Queue<i32> = Queue::new();
        assert!(queue.is_empty());
    }

    #[test]
    fn test_enqueue_dequeue() {
        let mut queue = Queue::new();
        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);

        assert_eq!(queue.dequeue(), Some(1));
        assert_eq!(queue.dequeue(), Some(2));
        assert_eq!(queue.dequeue(), Some(3));
        assert_eq!(queue.dequeue(), None);
    }

    #[test]
    fn test_peek() {
        let mut queue = Queue::new();
        assert_eq!(queue.peek(), None);

        queue.enqueue(42);
        assert_eq!(queue.peek(), Some(&42));
        assert_eq!(queue.len(), 1);
    }

    #[test]
    fn test_fifo_order() {
        let mut queue = Queue::new();
        for i in 0..5 {
            queue.enqueue(i);
        }
        for i in 0..5 {
            assert_eq!(queue.dequeue(), Some(i));
        }
    }
}
